crates/topos-tce-api/src/grpc/mod.rs

```diff
@@ -179,19 +183,71 @@ impl ApiService for TceGrpcService {
                             &subnet_id
                         )))
                     }
-                    Ok(Err(_)) => Err(Status::internal(
-                        "Can't get source head certificate position",
-                    )),
-                    Err(_) => Err(Status::internal(
-                        "Can't get source head certificate position",
-                    )),
+                    Ok(Err(e)) => Err(Status::internal(format!(
+                        "Can't get source head certificate position: {e}"
+                    ))),
+                    Err(e) => Err(Status::internal(format!(
+                        "Can't get source head certificate position: {e}"
+                    ))),
                 })
                 .await
         } else {
             Err(Status::invalid_argument("Certificate is malformed"))
         }
     }
 
+    async fn get_last_pending_certificates(
+        &self,
+        request: Request<GetLastPendingCertificatesRequest>,
+    ) -> Result<Response<GetLastPendingCertificatesResponse>, Status> {
+        let data = request.into_inner();
+
+        let subnet_ids = data.subnet_ids;
+
+        let (sender, receiver) = oneshot::channel();
+
+        let subnet_ids: Vec<SubnetId> = subnet_ids
+            .into_iter()
+            .map(TryInto::try_into)
+            .map(|v| v.map_err(|e| Status::internal(format!("Invalid subnet id: {e}"))))
+            .collect::<Result<_, _>>()?;
+
+        if self
+            .command_sender
+            .send(InternalRuntimeCommand::GetLastPendingCertificates { subnet_ids, sender })
+            .await
+            .is_err()
+        {
+            return Err(Status::internal(
+                "Can't get last pending certificates: sender dropped",
+            ));
+        }
+
+        receiver
+            .map(|value| match value {
+                Ok(Ok(map)) => Ok(Response::new(GetLastPendingCertificatesResponse {
+                    last_pending_certificate: map
+                        .into_iter()
+                        .map(|(subnet_id, cert)| {
+                            (
+                                subnet_id,
```

Comment:
You need to `base64` encode here following the discussion we had

Reply:
I don't think it applies here - this is not about debug display. It is about result - and we can not have `SubnetId` as key in the map so we have to use string representation. Both sequencer and tce are handling SubnetId as hex encoded string when dealing with them so I would keep hex encoded string here.

Reply:
If you use base64 encoded string here it will be again base64 encoded when grpc performs encoding/decoding

Reply:
In the `get_last_pending_certificates` response, the type of the key is `string` not `bytes` so there is no `base64` encoding by default. As example:

```
{
  "subnetIds": [
    {
      "value": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE="
    },
    {
      "value": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTI="
    }
  ]
}
```

This is a valid payload for the query with the `subnet_ids: [1111111111111111111111111111111, 11111111111111111111111111111112];`.

Here's the response with a dummy certificate on one subnet:

```
{
  "lastPendingCertificate": {
    "0x3131313131313131313131313131313131313131313131313131313131313131": {
      "value": {
        "prevId": {
          "value": "MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDA="
        },
        "sourceSubnetId": {
          "value": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE="
        },
        "stateRoot": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE=",
        "txRootHash": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE=",
        "targetSubnets": [
          {
            "value": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE="
          }
        ],
        "verifier": 507,
        "id": {
          "value": "NTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE="
        },
        "proof": {
          "value": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE="
        },
        "signature": {
          "value": "MTExMTExMTExMTExMTExMTExMTExMTExMTExMTExMTE="
        }
      }
    },
    "0x3131313131313131313131313131313131313131313131313131313131313132": {}
  }
}
```

As you can see, the `key` isn't encoded in base64, the `sourceSubnetid` is encoded tho. The main issue for me is that we have a lack of consistency here.

Reply:
Ok, I will change encoding to `base64` here